File: codingTest/kotlin/algorithm/src/one/블랙잭.kt
package one

fun main() {
    val (N, M) = readLine()!!.toString().split(' ').map {it.toInt()}
    val arr = readLine()!!.toString().split(' ').map { it.toInt() }
    if (arr.size != N) return
    println(nearSum(M, arr))
}

// 경우의 수
// C(n, 3) -> n개에서 3개를 뽑는 경우의 수 n*(n-1)*(n-2)/3!
fun nearSum(M: Int, arr: List<Int>): Int {
    var sum: Int = 0
    var result: Int = 0
    for (i in 0..arr.lastIndex) {
        for (j in (i+1)..arr.lastIndex) {
            for (k in (j+1)..arr.lastIndex) {
                sum = arr[i] + arr[j] + arr[k]
                if (sum <= M)  result = result.coerceAtLeast(sum)
            }
        }
    }
    return result
}

File: codingTest/javascript/블랙잭.js
var inputLines = require("fs").readFileSync("/dev/stdin").toString().split("\n");

async function main() {
  var [N, M] = inputLines[0].split(" ").map((e) => Number(e));
  var arr = inputLines[1].split(" ").map((e) => Number(e));
  if (arr.length !== N) return;
  var sum;
  var result = 0;
  for (var i = 0; i < arr.length; i++) {
    for (var j = i + 1; j < arr.length; j++) {
      for (var k = j + 1; k < arr.length; k++) {
        sum = arr[i] + arr[j] + arr[k];
        if (sum === M) return console.log(sum);
        if (result < sum && sum <= M) result = sum;
      }
    }
  }
  console.log(result);
}

main();
